src/lib.rs

```diff
@@ -308,165 +173,288 @@ impl KafkaJsonToDelta {
+                            latency_timer = Instant::now();
 
-                            let add = delta_writer.complete_file(&partition_cols).await?;
-                            let action = deltalake::action::Action::add(add);
+                            info!("Writing parquet file.");
+
+                            // Lookup the last transaction version from the Delta Log
+                            let last_txn_version = delta_writer
+                                .last_transaction_version(self.app_id.as_str())
+                                .await;
+
+                            // Create the new WAL entry
+                            let wal_entry = if let Ok(Some(txn_version)) = last_txn_version {
+                                info!(
+                                    "Delta table last txn version for {} is {}.",
+                                    self.app_id, txn_version
+                                );
+                                let last_wal_entry =
+                                    wal.get_entry_by_transaction_id(txn_version).await?;
+                                last_wal_entry.prepare_next(&partition_offsets)
+                            } else {
+                                info!("Delta table does not contain a txn for app id {}. Starting write-ahead-log from 1.", self.app_id);
+                                WriteAheadLogEntry::new(
+                                    1,
+                                    TransactionState::Prepared,
+                                    None, /*delta table version starts out empty*/
+                                    partition_offsets.clone(),
+                                )
+                            };
+                            wal.put_entry(&wal_entry).await?;
+
+                            // Complete the file
+                            let txn = action::Action::txn(action::Txn {
+                                appId: self.app_id.clone(),
+                                version: wal_entry.transaction_id,
+                                lastUpdated: std::time::SystemTime::now()
+                                    .duration_since(std::time::UNIX_EPOCH)
+                                    .unwrap()
+                                    .as_millis()
+                                    as i64,
+                            });
+                            let commit_result = delta_writer.write_file(&mut vec![txn]).await;
+
+                            match commit_result {
+                                Ok(version) => {
+                                    wal.complete_entry(wal_entry.transaction_id, version)
+                                        .await?;
+                                }
+                                Err(e) => {
+                                    error!("Delta commit failed. Aborting write ahead log entry.");
+                                    wal.abort_entry(wal_entry.transaction_id).await?;
```

Comment:
Another killer `?` perhaps?

Reply:
Yes - but simpler than the others I think. This one can be solved with retry loop, but without worrying much about other consumers. Giving my same excuse here anyway.